### PR TITLE
fix: fix #720 Change type inference for function parameters

### LINKED ISSUE #720: Default type for an assignment patterns should respect all values
https://github.com/dudykr/stc/blob/1cb6c1844cfd2e6b069bddc3588dda51d53a3b6f/crates/stc_ts_type_checker/tests/conformance/es6/destructuring/destructuringWithLiteralInitializers2.ts#L15

Currently `stc` infers the type as `number[]`, but it should be `[number, (string | undefined)]`

### PR DESCRIPTION
**Pull request by blacktoast**
**Description:**
fix #720 

Fixed issue #720 by changing the `add_vars` function.

**BREAKING CHANGE:**

**Related issue (if exists):**
#720

### CODE REVIEW

File: crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs [hunk 1]
```diff
@@ -357,9 +363,134 @@ impl Analyzer<'_, '_> {
...
+                                                                    }
+                                                                }
+
+                                                                if !has_undefined {
+                                                                    union.types.push(Type::undefined(span, Default::default()));
+                                                                }
+
+                                                                box Type::Union(union)
+                                                            }
+                                                            _ => match elem_ty.clone().keyword().as_ref().unwrap().kind {
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs:411]: Please use pattern matching like 

```rust
match elem_ty.normalize() {
    Type::Keyword(KeywordType { kind: ... }) => { ... }
}
```
instead

File: crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs [hunk 2]
```diff
@@ -357,9 +363,134 @@ impl Analyzer<'_, '_> {
...
+                                                    if is_not_assigned_type {
+                                                        let elem_ty = p
+                                                            .right
+                                                            .as_ref()
+                                                            .validate_with_default(self)?
+                                                            .clone()
+                                                            .fold_with(&mut Widen { tuple_to_array: false });
+
+                                                        let convert_ty = match elem_ty {
+                                                            Type::Union(mut union) => {
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs:394]: Please use a different name than `union`. It disturbs code review

File: crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs [hunk 3]
```diff
@@ -357,9 +363,134 @@ impl Analyzer<'_, '_> {
...
+                                                RPat::Assign(p) => {
+                                                    if is_not_assigned_type {
+                                                        let elem_ty = p
+                                                            .right
+                                                            .as_ref()
+                                                            .validate_with_default(self)?
+                                                            .clone()
+                                                            .fold_with(&mut Widen { tuple_to_array: false });
+
+                                                        let convert_ty = match elem_ty {
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs:393]: Please match on `elem_ty.as_union_type_mut()` instead

File: crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs [hunk 4]
```diff
@@ -523,7 +654,7 @@ impl Analyzer<'_, '_> {
                     }
 
                     real_ty.freeze();
-
+                    println!("real_ty {:#?}", real_ty);
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs:657]: Please remove this line

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thanks!

File: crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs [hunk 5]
```diff
@@ -357,9 +363,144 @@ impl Analyzer<'_, '_> {
...
+                                        let is_not_assigned_type = i > right_type_len - 1;
+                                        if let Some(r_pat) = left_element {
+                                            match r_pat {
+                                                RPat::Assign(p) => {
+                                                    if is_not_assigned_type {
+                                                        let mut elem_ty = p
+                                                            .right
+                                                            .as_ref()
+                                                            .validate_with_default(self)?
+                                                            .clone()
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs:392]: This clone seems redundant

Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs:392]: But well, I think we can defer this

File: crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs [hunk 6]
```diff
@@ -357,9 +363,144 @@ impl Analyzer<'_, '_> {
...
                         }
                     }
+                    // Type inference for functions
+                    let default_ty = match default_ty {
+                        Some(d_ty) => {
+                            let d_ty = d_ty.fold_with(&mut Widen { tuple_to_array: false });
+
+                            match d_ty {
+                                Type::Tuple(mut ty) => {
+                                    let right_type_len = match ty.elems.len() {
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs:377]: This is wrong

File: crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs [hunk 7]
```diff
@@ -357,9 +363,144 @@ impl Analyzer<'_, '_> {
...
+                                        0 => 1,
+                                        _ => ty.elems.len(),
+                                    };
+
+                                    for (i, left_element) in arr.elems.iter().enumerate() {
+                                        let is_not_assigned_type = i > right_type_len - 1;
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs:383]: You should do `i + 1> right_type_len` instead

File: crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs (hunk 5 above)
Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs:392]: Please fix this, too.

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs (hunk 7 above)
Comment (blacktoast) [reply crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs:383]: Oh, that makes sense

**Comment by blacktoast** [comment]
I cleared it and pushed it again

Thanks for the code review

**Review by kdy1 — APPROVED** [review]
Thank you!

**Pull request merged**